Test against RFC4291 for IPv6 address strings [10190]

RFRIEDM opened the pull request:
This PR is add tests for IPv6 address representation. Adhering to [RFC4291](https://tools.ietf.org/html/rfc4291#section-2.3) will test for some issues found in #1613. 
The underlying issues still need to be fixed; this PR only addresses the lack of test coverage for them.

* Test zero compression
* Test handling uppercase/lowercase
* Test equivalance of toIPv6string overrides
* Add utility for resetting a locator's address to default

Signed-off-by: Ryan Friedman <ryan_friedman@trimble.com>

richiware commented:
Build status:
* Linux [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux&build=3391)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3391/)
* Linux aarch64[![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux%20aarch64&build=589)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/589/)
* Mac [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Mac&build=3041)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3041/)
* Windows [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Windows&build=3423)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3423/)

EduPonz commented:
Hi @RFRIEDM ,

Could you please uncrustify your changes? Do you plan to address the fixes as well, or should we take care of them?

RFRIEDM commented:
> Hi @RFRIEDM ,
> 
> Could you please uncrustify your changes? Do you plan to address the fixes as well, or should we take care of them?

Yes - can you direct me to where the uncrustify config file is mentioned in the quality guidelines? 

If I don't see the bug fixed by the end of the year, I'll take a stab at it, but in the mean time, was planning to try out whitelisting interfaces with IPv6.

EduPonz commented:
Hi @RFRIEDM,

You can find the contributing guidelines in CONTRIBUTING.md. There, there is a link to the uncrustify config file.

As for the fix, we won't have time to do it before next year, since most of us are already on vacation. We can take a look at it after January the 7th, but feel free to take a crack at it if you want.

Whitelisting IPv6 interfaces should be working, in fact it's the very reason we reworked this part. In any case, contributions in the form of either PRs or issue reports are most welcomed, so thanks for your time and help!

RFRIEDM commented:
> Hi @RFRIEDM,
> 
> You can find the contributing guidelines in CONTRIBUTING.md. There, there is a link to the uncrustify config file.
> 
> As for the fix, we won't have time to do it before next year, since most of us are already on vacation. We can take a look at it after January the 7th, but feel free to take a crack at it if you want.
> 
> Whitelisting IPv6 interfaces should be working, in fact it's the very reason we reworked this part. In any case, contributions in the form of either PRs or issue reports are most welcomed, so thanks for your time and help!

Excellent - I force pushed the uncrustified file and it looks like it passed that. Just waiting for jenkins to finish. 

I implemented a fix for this issue today. Once this PR is merged, I'll submit a PR for the fix. 

Thank you for all the help. I look forward to proving out whitelisting and then getting all that running on some Raspberry Pi 4's.

EduPonz commented:
Hi @RFRIEDM,

I've seen you've the fix already implemented, that's fast! I don't think we want to merge tests that we know are going to fail until the fix arrives. However, I think what we can do is to review your tests and give you feedback. Once the tests are reviewed, you could push the fix to this same PR to have a final review before merging. That way, we make sure that we test everything that needs to be tested, and at the same time we don't break CI builds in the process. Does this sound good to you?

@imontesino, could you take a look at these changes?

RFRIEDM commented:
> Hi @RFRIEDM,
> 
> I've seen you've the fix already implemented, that's fast! I don't think we want to merge tests that we know are going to fail until the fix arrives. However, I think what we can do is to review your tests and give you feedback. Once the tests are reviewed, you could push the fix to this same PR to have a final review before merging. That way, we make sure that we test everything that needs to be tested, and at the same time we don't break CI builds in the process. Does this sound good to you?
> 
> @imontesino, could you take a look at these changes?

Yes, the fix with a clean commit history can be found on a separate branch [here](https://github.com/RFRIEDM/Fast-DDS/commit/d174703168df186a64b39def774354714b1f323f) that I originally intended to submit as a separate PR.

Your alternative proposal sounds good.

I did go down a bit of a rabbit hole with the socket library and had a few concerns with merging the fix. 
1. I only ran it on Linux - there's a chance I added a dependency on some define in the socket library that may arise in CI
2. It adds includes that the team may not want added of `netdb.h` and `arpa/inet.h`
3. It touches more code than just fixing the bug in an effort to make the code more in alignment with the RFC's for IPv6
4. So far, I only addressed IPv6. If you do approve the fix, I would be happy to finish it out for IPv4 (a simpler task)

If you don't want added dependencies, [here's](https://gist.github.com/njh/84125c8ededdeb74ec5cc80a4003f308) an alternative resource to fix toIPv6string that can be used without dependencies. 

-Ryan

EduPonz commented:
Hi @RFRIEDM ,

I haven't thoroughly looked at your fix cause I do not want that to interfere with the test revision (which should come first since it's a requirement). I do however have a few comments:

> * It adds includes that the team may not want added of `netdb.h` and `arpa/inet.h`

It looks like this dependencies might have problems in non-POSIX platforms, and we do officially support for Windows 10 VS2019 of both 32 and 64 bits. We also run CI on VS2017, since we won't like to break that compatibility. For socket operations we use [Boost's Asio](https://think-async.com/Asio/), which is cross-platform and already a dependency of Fast DDS. Maybe it's worth looking at it to subtitute glib (from a very qucik search on API ref [make_address_v6](https://think-async.com/Asio/boost_asio_1_18_0/doc/html/boost_asio/reference/ip__address_v6/make_address_v6.html))

> * It touches more code than just fixing the bug in an effort to make the code more in alignment with the RFC's for IPv6

Every improvement in performance, readability, and of course standars adherence is deeply appreciated. I see this as an upside rather than a concern. 

> * So far, I only addressed IPv6. If you do approve the fix, I would be happy to finish it out for IPv4 (a simpler task)

That'd be amazing!
Thanks for all the iterations!

EduPonz commented:
I've marked the PR as `no-test` to save CI resources until we have a commit with the fix.

imontesino commented on test/unittest/transport/UDPv6Tests.cpp:
```diff
@@ -89,15 +103,129 @@ TEST_F(UDPv6Tests, conversion_to_ip6_string)
+    ASSERT_EQ("::1000", IPLocator::toIPv6string(locator));
+    locator.address[15] = 0x20;
+    ASSERT_EQ("::1020", IPLocator::toIPv6string(locator));
+
+    //Embedded Zeros in 2001:DB8:a::/48
+    reset_locator_address(locator);
```
Partial review.
I have reviewed the tests, they are very thorough and cover the all the cases of address compression. A nitpick/question, is this `/48` a typo?

Feel free to commit the fix when you want, and I will continue the review.

RFRIEDM replied:
Yes copy and paste from the RFC that is not relevant. Will fix

richiware commented:
Build status:
* Linux [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux&build=3416)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3416/)
* Linux aarch64[![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux%20aarch64&build=610)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/610/)
* Mac [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Mac&build=3054)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3054/)
* Windows [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Windows&build=3448)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3448/)

RFRIEDM commented:
> Hi @RFRIEDM ,
> 
> I haven't thoroughly looked at your fix cause I do not want that to interfere with the test revision (which should come first since it's a requirement). I do however have a few comments:
> 
> > * It adds includes that the team may not want added of `netdb.h` and `arpa/inet.h`
> 
> It looks like this dependencies might have problems in non-POSIX platforms, and we do officially support for Windows 10 VS2019 of both 32 and 64 bits. We also run CI on VS2017, since we won't like to break that compatibility. For socket operations we use [Boost's Asio](https://think-async.com/Asio/), which is cross-platform and already a dependency of Fast DDS. Maybe it's worth looking at it to subtitute glib (from a very qucik search on API ref [make_address_v6](https://think-async.com/Asio/boost_asio_1_18_0/doc/html/boost_asio/reference/ip__address_v6/make_address_v6.html))
> 
Thank you for the references I will look into them. 

> > * It touches more code than just fixing the bug in an effort to make the code more in alignment with the RFC's for IPv6
> 
> Every improvement in performance, readability, and of course standars adherence is deeply appreciated. I see this as an upside rather than a concern.
> 

👍 
> > * So far, I only addressed IPv6. If you do approve the fix, I would be happy to finish it out for IPv4 (a simpler task)
> 
> That'd be amazing!
> Thanks for all the iterations!

Of course!

richiware commented:
Build status:
* Linux [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux&build=3439)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3439/)
* Linux aarch64[![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Linux%20aarch64&build=632)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/632/)
* Mac [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Mac&build=3064)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3064/)
* Windows [![Build Status](http://jenkins.eprosima.com:8080/buildStatus/icon?job=FastRTPS%20Manual%20Windows&build=3473)](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3473/)

jparisu commented:
Hi @RFRIEDM ,

First of all, thank you for your time and contribution.

I've been checking your PR and I have a question that has been chasing me for a while.
The conversion from IPv6 to string is not unique, and nor in the RFC, nor in any webpage, I have been able to find a recommendation on how to compress it.

In our opinion, there are three logical solutions:
1. following the RFC 4291 recommendation and not compress at all. (Too large in some cases as :: or ::1)
2. compress the first zero block in the IP. (Faster but not the most efficient compression)
3. compress the larger zero block in the IP. (Slower but with efficiently compression)

Considering that the three of them are equally correct, and all of them could be unequivocally reconverted to a IPv6, we have chosen the second one, as it is faster between the two compression options.

Looking at your PR, I understand that your problem came from a bug in the `toIPv6string` function.
Also, the new test cases that you have added are fulfilled with this fix except to the case commented before. I have changed `2001:db8:a:0:1::` to `2001:db8:a::1:0:0:0` in order to follow out approach. These tests have been changed to `LocatorTests.cpp` in order to unify the Locator functions in the same file (if they were added in UDPv6 they should be added as well in TCPv6).

The new PR with this fix and the new test cases is #1671 .

RFRIEDM commented:
> Hi @RFRIEDM ,
> 
> First of all, thank you for your time and contribution.
> 
> I've been checking your PR and I have a question that has been chasing me for a while.
> The conversion from IPv6 to string is not unique, and nor in the RFC, nor in any webpage, I have been able to find a recommendation on how to compress it.
> 
> In our opinion, there are three logical solutions:
> 
>     1. following the RFC 4291 recommendation and not compress at all. (Too large in some cases as :: or ::1)
> 
>     2. compress the first zero block in the IP. (Faster but not the most efficient compression)
> 
>     3. compress the larger zero block in the IP. (Slower but with efficiently compression)
> 
> 
> Considering that the three of them are equally correct, and all of them could be unequivocally reconverted to a IPv6, we have chosen the second one, as it is faster between the two compression options.
> 
> Looking at your PR, I understand that your problem came from a bug in the `toIPv6string` function.
> Also, the new test cases that you have added are fulfilled with this fix except to the case commented before. I have changed `2001:db8:a:0:1::` to `2001:db8:a::1:0:0:0` in order to follow out approach. These tests have been changed to `LocatorTests.cpp` in order to unify the Locator functions in the same file (if they were added in UDPv6 they should be added as well in TCPv6).
> 
> The new PR with this fix and the new test cases is #1671 .

Sorry for they delay in response. Yes, you are correct that RFC does not specify which block of zero's to compress, however option 3 appears to be commonly implemented. If there are equal length blocks of zeros, it also appears normal to compress the first block. Looking around further, I found [RFC5952 section 4.2.3](https://tools.ietf.org/html/rfc5952#section-4.2.3) affirms the behaviour. Given it's in the RFC to follow option 3, I propose to adhere to that too.

* [ultratools.com](https://www.ultratools.com/tools/ipv6CompressResult?address=2001%3Adb8%3A0%3A0%3A200%3A0%3A0%3A7) compresses the first largest block
* [ipaddressguide](https://www.ipaddressguide.com/ipv6) compresses the first largest block
* * [findipv6](https://findipv6.com/ipv6-compress/) compresses the first largest block 
* [Geeks for geeks](https://www.geeksforgeeks.org/compression-of-ipv6-address/) says that compressing either equal block is ok

jparisu commented:
Hi @RFRIEDM ,

Again, thank you for your contribution.
We have applied RFC 5952 recommendations and use the compression for the first largest block.

Please, feel free to check the implementation in the PR #1685 and add more comments or reviews.

RFRIEDM commented:
#1685 Looks good, so closing this PR as a no-merge and the associated issue.
